columnflow/tasks/coalesce.py

```diff
@@ -0,0 +1,152 @@
+        shifts |= cls.dep_ProduceColumns.get_allowed_shifts(config_inst, params)
+        return shifts
+
+    def workflow_requires(self, only_super: bool = False):
+        reqs = super().workflow_requires()
+        if only_super:
+            return reqs
+
+        # require the full merge forest
+        reqs["events"] = self.dep_MergeReducedEvents.req(self, tree_index=-1)
+
+        if not self.pilot:
+            if self.producers:
+                reqs["producers"] = [
+                    self.dep_ProduceColumns.req(self, producer=p)
+                    for p in self.producers
+                ]
+            if self.ml_models:
+                reqs["ml"] = [
+                    self.dep_MLEvaluation.req(self, ml_model=m)
+                    for m in self.ml_models
+                ]
+
+        return reqs
+
+    def requires(self):
+        reqs = {
+            "events": self.dep_MergeReducedEvents.req(self, tree_index=self.branch, _exclude={"branch"}),
+        }
+
+        if self.producers:
+            reqs["producers"] = [
+                self.dep_ProduceColumns.req(self, producer=p)
+                for p in self.producers
+            ]
+        if self.ml_models:
+            reqs["ml"] = [
+                self.dep_MLEvaluation.req(self, ml_model=m)
+                for m in self.ml_models
+            ]
+
+        return reqs
+
+    @MergeReducedEventsUser.maybe_dummy
+    def output(self):
+        return self.target(f"data_{self.branch}.parquet")
+
+    @law.decorator.log
+    @law.decorator.localize(input=True, output=False)
+    @law.decorator.safe_output
+    def run(self):
+        from columnflow.columnar_util import RouteFilter, update_ak_array, sorted_ak_to_parquet
+
+        # prepare inputs and outputs
+        inputs = self.input()
+        output = self.output()
+        output_chunks = {}
+
+        # create a temp dir for saving intermediate files
+        tmp_dir = law.LocalDirectoryTarget(is_tmp=True)
+        tmp_dir.touch()
+
+        # define nano columns that should be kept, and that need to be loaded
+        keep_columns = set(self.config_inst.x.keep_columns[self.task_family])
+        # load_columns = keep_columns | set(mandatory_coffea_columns)
+        route_filter = RouteFilter(keep_columns)
+
+        # iterate over chunks of events and diffs
+        files = [inputs["events"]["collection"][0].path]
+        if self.producers:
+            files.extend([inp.path for inp in inputs["producers"]])
+        if self.ml_models:
+            files.extend([inp.path for inp in inputs["ml"]])
+        for (events, *columns), pos in self.iter_chunked_io(
+            files,
+            source_type=len(files) * ["awkward_parquet"],
+            # TODO: not working yet since parquet columns are nested
+            # open_options=[{"columns": load_columns}] + (len(files) - 1) * [None],
+        ):
+            # add additional columns
+            events = update_ak_array(events, *columns)
```

Comment:
Just to clarify: if the same column exist in multiple files, the priority is:

MLEvaluation > ProduceColumns > ReduceEvents ?